provider/coredns.go

```diff
@@ -83,52 +90,56 @@ type Service struct {
 }
 
 type etcdClient struct {
-	api etcd.KeysAPI
+	client *etcdcv3.Client
+	ctx    context.Context
 }
 
-var _ skyDNSClient = etcdClient{}
+var _ coreDNSClient = etcdClient{}
 
 // GetService return all Service records stored in etcd stored anywhere under the given key (recursively)
 func (c etcdClient) GetServices(prefix string) ([]*Service, error) {
-	var result []*Service
-	opts := &etcd.GetOptions{Recursive: true}
-	data, err := c.api.Get(context.Background(), prefix, opts)
+	ctx, cancel := context.WithTimeout(c.ctx, etcdTimeout)
+	defer cancel()
+
+	path := prefix
+	r, err := c.client.Get(ctx, path, etcdcv3.WithPrefix())
 	if err != nil {
-		if etcd.IsKeyNotFound(err) {
-			return nil, nil
-		}
 		return nil, err
 	}
 
-	queue := list.New()
-	queue.PushFront(data.Node)
-	for queueNode := queue.Front(); queueNode != nil; queueNode = queueNode.Next() {
-		node := queueNode.Value.(*etcd.Node)
-		if node.Dir {
-			for _, childNode := range node.Nodes {
-				queue.PushBack(childNode)
-			}
-			continue
+	var sx []*Service
+	bx := make(map[Service]bool)
```

Comment:
For someone which reads the code it's hard to tell what it does, could you find a better name? or just comment what it does :)

Reply:
I have added comments now. sorry for the issue in readability. i referred to the `etcd` plugin in `coredns` [here](https://github.com/coredns/coredns/blob/master/plugin/etcd/etcd.go#L144:L148) and didn't do to much of my own modifications.